src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsParserStateObject.cs

```diff
@@ -488,6 +489,19 @@ internal bool TryStartNewRow(bool isNullCompressed, int nullBitmapColumnsCount =
         internal bool TryReadChars(char[] chars, int charsOffset, int charsCount, out int charsCopied)
         {
             charsCopied = 0;
+            if (!BitConverter.IsLittleEndian)
+            {
+                charsCopied = charsCount;
+
+                for (int ii = 0; ii < charsCopied; ii++)
+                {
+                    if (!TryReadChar(out chars[charsOffset + ii]))
+                    {
+                        return false;
+                    }
+                }
+                return true;
+            }
```

Comment:
isn't this going to cause a performance problem? in the BE case it'll do char by char reads rather than copying in chunks. It's be better to read the data completely and then do a byteswap just before converting to a string, otherwise any position where we're off-by-one in bytes would do an incorrect swap.